src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,532 @@
+            }
+        }
+
+        private void WriteSyntaxTree(
+            JsonWriter writer,
+            SyntaxTreeKey syntaxTree,
+            ImmutableArray<KeyValuePair<string, string>> pathMap,
+            DeterministicKeyOptions options,
+            CancellationToken cancellationToken)
+        {
+            writer.WriteObjectStart();
+            WriteFilePath(writer, "fileName", syntaxTree.FilePath, pathMap, options);
+            writer.WriteKey("text");
+            WriteSourceText(writer, syntaxTree.GetText(cancellationToken));
+            writer.WriteKey("parseOptions");
+            WriteParseOptions(writer, syntaxTree.Options);
+            writer.WriteObjectEnd();
+        }
+
+        private void WriteSourceText(JsonWriter writer, SourceText? sourceText)
+        {
+            if (sourceText is null)
+            {
+                return;
+            }
+
+            writer.WriteObjectStart();
+            WriteByteArrayValue(writer, "checksum", sourceText.GetChecksum().AsSpan());
+            writer.Write("checksumAlgorithm", sourceText.ChecksumAlgorithm);
+            writer.Write("encodingName", sourceText.Encoding?.EncodingName);
+            writer.WriteObjectEnd();
+        }
+
+        internal void WriteMetadataReference(
+            JsonWriter writer,
+            MetadataReference reference,
+            ImmutableArray<KeyValuePair<string, string>> pathMap,
+            DeterministicKeyOptions deterministicKeyOptions,
+            CancellationToken cancellationToken)
+        {
+            writer.WriteObjectStart();
+            if (reference is PortableExecutableReference peReference)
+            {
+                switch (peReference.GetMetadata())
+                {
+                    case AssemblyMetadata assemblyMetadata:
+                        {
+                            var modules = assemblyMetadata.GetModules();
+                            writeModuleMetadata(modules[0]);
+                            writer.WriteKey("secondaryModules");
+                            writer.WriteArrayStart();
+                            for (var i = 1; i < modules.Length; i++)
+                            {
+                                writer.WriteObjectStart();
+                                writeModuleMetadata(modules[i]);
+                                writer.WriteObjectEnd();
+                            }
+                            writer.WriteArrayEnd();
+                        }
+                        break;
+                    case ModuleMetadata m:
+                        writeModuleMetadata(m);
+                        break;
+                    case var m:
+                        throw ExceptionUtilities.UnexpectedValue(m);
+                }
+
+                writer.WriteKey("properties");
+                writeMetadataReferenceProperties(writer, reference.Properties);
+
+                void writeModuleMetadata(ModuleMetadata moduleMetadata)
```

Comment:
nit: We try to put all local functions at the end of the method, if possible, to avoid interrupting the main flow of the code